src/myst/tools/repl.cr

```diff
@@ -0,0 +1,75 @@
+require "readline"
+
+module Myst
+  class Repl
+
+    def self.start
+      new.start
+    end
+
+    def initialize
+      @history = IO::Memory.new
+      @interpreter = Interpreter.new
+    end
+
+    def process(level, prompt, prev_input = "")
+      loop do
+        if input = wait_for_input(prompt)
+          new_input = "#{prev_input}\n#{input}"
+          parser = Parser.new(IO::Memory.new(new_input), "")
+
+          begin
+            program = parser.parse
+            @interpreter.run(program)
+
+            STDOUT.print "=> "
+            STDOUT.flush
+            STDOUT.puts stack_value(@interpreter.stack.pop)
+            return unless level == 0                          
+          rescue ex 
+            if ex.message.to_s.includes?("EOF")
```

Comment:
Something I though about a lot when I did the parser rewrite back in September was making `Parser` be streamable, where the parser (really the lexer) will just block and wait to read the next token of input if one isn't available yet.

This seems like a good use case for trying to use a stream of input. My thought would be a custom descendant of `IO` that is responsible for printing the prompt and reading in input. The lexer will then ask that class for more input when it needs it (through a [proxy `Reader` class](https://github.com/myst-lang/myst/blob/master/src/myst/syntax/reader.cr#L15), and if the input stream runs out of input to give back, it can render a new prompt to get more (similar to how the `level` part is working here).

`Parser#parse` will then return when it has received enough input to complete an expression, or it will raise an error when it encounters one.

I don't know how much work that would be, but I think that would be a bit nicer than checking error output for `EOF` after every line.

Reply:
I agree, this is really just a hack to get it working.  I originally tried to stream the Parser, and maybe we can revisit it